tools/codegen/api/unboxing.py

```diff
@@ -0,0 +1,224 @@
+    if isinstance(t, BaseType):
+        out_name = f"{arg_name}_base"
+        code = [f"{ctype} {arg_name}_base = {ival}.to<{ctype}>();"]
+    elif isinstance(t, OptionalType):
+        in_name = arg_name + "_opt_in"
+        out_name = arg_name + "_opt_out"
+        connector = "\n\t"
+        res_expr, res_code = argumenttype_ivalue_convert(t.elem, in_name, arg_name)
+        code = f"""
+c10::optional<c10::IValue> {arg_name + "_opt"} = {ival}.toOptional<c10::IValue>();
+{ctype} {out_name};
+if ({arg_name + "_opt"}.has_value()) {{
+    const c10::IValue {in_name} = {arg_name + "_opt"}.value();
+    {connector.join(res_code)}
+    {out_name} = {ctype}({res_expr.expr});
+}} else {{
+    {out_name} = {ctype}();
+}}
+        """.split(
+            "\n"
+        )
+    elif isinstance(t, ListType):
+        in_name = arg_name + "_list_in"
+        out_name = arg_name + "_list_out"
+        elem_name = arg_name + "_elem"
+        code = [f"const c10::List<c10::IValue> {in_name} = {ival}.toList();"]
+        res_expr, res_code = argumenttype_ivalue_convert(t.elem, elem_name, arg_name)
+        connector = "\n\t"
+        # handle list type with size, e.g., bool[4]
+        if isinstance(t.elem, BaseType) and t.elem.name == BaseTy.bool and t.size:
+            code.extend(
+                f"""
+{ctype} {out_name} = as_array<{res_expr.type.cpp_type(strip_ref=True)}, {t.size}>({in_name});
+            """.split(
+                    "\n"
+                )
+            )
+        # we have to use c10::List for optional element. e.g., Tensor?[] -> c10::List<c10::optional<at::Tensor>>
+        elif isinstance(t.elem, OptionalType):
+            code.extend(
+                f"""
+{ctype} {out_name};
+for (c10::IValue {elem_name}: {in_name}) {{
+    {connector.join(res_code)}
+    {out_name}.push_back({res_expr.expr});
+}}
+            """.split(
+                    "\n"
+                )
+            )
+        else:
+            # use ArrayRef as default.
+            vec_name = arg_name + "_vec"
+            code.extend(
+                f"""
+std::vector<{res_expr.type.cpp_type(strip_ref=True)}> {vec_name};
+for (c10::IValue {elem_name}: {in_name}) {{
+    {connector.join(res_code)}
+    {vec_name}.push_back({res_expr.expr});
+}}
+{ctype} {out_name}({vec_name});
+            """.split(
+                    "\n"
+                )
+            )
+    else:
+        raise Exception(f"Cannot handle type {t}. ival: {ival}, arg_name: {arg_name}")
+    expr = Expr(expr=out_name, type=nctype)
+    return expr, code
+
+
+# Generate code to call C++ unboxed kernel with argument variable names
+def generate_unboxed_kernel_call(
+        f: NativeFunction, sig: CppSignature, arguments: List[Expr]
+) -> List[str]:
+    use_tensor_options = any(
+        isinstance(a.nctype.type, BaseCType) and a.nctype.type.type == tensorOptionsT
+        for a in sig.arguments()
+    )
+    # find dispatch native function namespace
+    if use_tensor_options:
```

Comment:
i still think we can get away without this: is there anything in the `torch` namespace that you actually need? You should be able to get away with using the `at::` namespace for everything here (since the output file also lives in `aten`)